sql/schema.go

```diff
@@ -0,0 +1,202 @@
+	var (
+		err   error
+		ignRx *regexp.Regexp
+		sb    strings.Builder
+	)
+	if ignoreRx != "" {
+		ignRx, err = regexp.Compile(ignoreRx)
+		if err != nil {
+			return "", fmt.Errorf("error compiling table ignore regexp %q: %w",
+				ignoreRx, err)
+		}
+	}
+	version, err := version(db)
+	if err != nil {
+		return "", err
+	}
+	fmt.Fprintf(&sb, "PRAGMA user_version = %d;\n", version)
+	if _, err = db.Exec(
+		// Type is either 'index' or 'table', we want tables
+		// to go first. Also, we ignore _litestream tables
+		`select tbl_name, sql || ';' from sqlite_master
+                 where sql is not null
+                 order by tbl_name, type desc, name`,
+		nil, func(st *Statement) bool {
+			if ignRx == nil || !ignRx.MatchString(st.ColumnText(0)) {
+				fmt.Fprintln(&sb, st.ColumnText(1))
+			}
+			return true
+		}); err != nil {
+		return "", fmt.Errorf("error retrieving DB schema: %w", err)
+	}
+	return sb.String(), nil
+}
+
+// Schema represents database schema.
+type Schema struct {
+	Script        string
+	Migrations    MigrationList
+	skipMigration map[int]struct{}
+}
+
+// Diff diffs the database schema against the actual schema.
+// If there's no differences, it returns an empty string.
+func (s *Schema) Diff(actualScript string) string {
+	if s.Script == actualScript {
+		return ""
+	}
+	diff := godiffpatch.GeneratePatch(SchemaPath, s.Script, actualScript)
+	if diff == "" {
+		return "<diff>"
+	}
+	return diff
+}
+
+// WriteToFile writes the schema to the corresponding updated schema file.
+func (s *Schema) WriteToFile(basedir string) error {
+	path := filepath.Join(basedir, UpdatedSchemaPath)
+	if err := os.WriteFile(path, []byte(s.Script), 0o777); err != nil {
+		return fmt.Errorf("error writing schema file %s: %w", path, err)
+	}
+	return nil
+}
+
+// SkipMigrations skips the specified migrations.
+func (s *Schema) SkipMigrations(i ...int) {
+	if s.skipMigration == nil {
+		s.skipMigration = make(map[int]struct{})
+	}
+	for _, index := range i {
+		s.skipMigration[index] = struct{}{}
+	}
+}
+
+// Apply applies the schema to the database.
+func (s *Schema) Apply(db *Database) error {
+	return db.WithTx(context.Background(), func(tx *Tx) error {
+		scanner := bufio.NewScanner(strings.NewReader(s.Script))
+		scanner.Split(func(data []byte, atEOF bool) (advance int, token []byte, err error) {
+			if i := bytes.Index(data, []byte(";")); i >= 0 {
+				return i + 1, data[0 : i+1], nil
+			}
+			return 0, nil, nil
+		})
+		for scanner.Scan() {
+			if _, err := tx.Exec(scanner.Text(), nil, nil); err != nil {
+				return fmt.Errorf("exec %s: %w", scanner.Text(), err)
+			}
+		}
+		return nil
+	})
+}
+
+// Migrate performs database migration. In case if migrations are disabled, the database
+// version is checked but no migrations are run, and if the database is too old and
+// migrations are disabled, an error is returned.
+func (s *Schema) Migrate(logger *zap.Logger, db *Database, vacuumState int, enable bool) error {
```

Comment:
I'm usually not a fan of passing bool operators like this. `Migrate` should in my opinion not change its behavior if `config.enableMigrations` is set to `true` or `false`. The caller should check if running migrations is desired or not and either fail directly or call `Migrate` if needed, rather than putting that logic into the `Migrate` method.

Reply:
`Migrate` still checks db version in both cases, whether the migrations are enabled or disabled. In case if they're enabled, the old db version causes the migrations to be run, and if they're disabled, an error is returned. Maybe this method needs to be renamed... Splitting it in two may not be very pretty (version comparison logic not done in Schema? Or `enable` still passed to one of methods split out?)

Reply:
It's about separation of concerns. In my opinion The caller of `Migrate` should check if a migration is necessary of not and if it is call `Migrate` instead of always calling `Migrate` and migrate determining if migration is necessary.

For me it becomes more logical the following way:

- check current version and determine target version
- if those are not equal and `config.enableMigrations` is false return an error
  - otherwise call migrate - its already a no-op if `before` and `after` are the same.

Reply:
Split `Migrate` into `CheckDBVersion` and `Migrate`